Fix host code appearing in Wasm binaries

JayAndJef opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

This is a direct fix for issue [132802](https://github.com/rust-lang/rust/issues/132802).
Followed the outline as follows:
> * give a hard error in bootstrap when using gcc to compile for wasm
> * change our CI to use clang instead of gcc
> * add a test that compiling a sample program for wasm32-unknown doesn't give any linker warnings

The `test-various` ci job was also changed.

try-job: test-various
try-job: dist-various-1
try-job: dist-various-2
try-job: x86_64-msvc-1

JayAndJef committed:
fix: wasm-bare targets compiling x86 builtins

Added sanity check to bootstrap to hard error on wasm builds without
clang, and changed distribution image `dist-various-2` to use clang to
build for official targets.

jyn514 commented on src/bootstrap/src/core/sanity.rs:
```diff
@@ -376,3 +387,8 @@ $ pacman -R cmake && pacman -S mingw-w64-x86_64-cmake
+
+fn is_gcc_compiler(path: PathBuf, build: &Build) -> bool {
+    let cc_output = command(&path).arg("--version").run_capture_stdout(build).stdout();
+    cc_output.contains("GCC")
```
i don't think this does what you want. on my machine, `cc --version` prints "cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0", even though it's really gcc.

i am not sure how to check this ... maybe the rust equivalent of `realpath $(which cc) | rg gcc`? or we could check `cc.is_like_clang()`: https://docs.rs/cc/latest/cc/struct.Tool.html#method.is_like_clang

JayAndJef replied:
I'll try to find a gcc-specific flag to run. If all else fails, we can try compiling a C program and check if some headers are defined.

alexcrichton replied:
When going down this path I might recommend something like `clang -E -dM -x c /dev/null | rg -i clang` which will print `#define __clang__ 1`. That's a test for clang rather than a test for "not gcc" but at this time only clang has support for wasm, so it might work as a detection mechanism nonetheless?

Although IMO rejecting compilers like this is a bit of a tricky business that's bound to have false positives and negatives so personally I'd probably just skip this part and rely on the test you added instead.

tgross35 commented on tests/run-make/wasm-builtins-no-linker-warning/rmake.rs:
This can just be a UI test (`tests/ui`), which are cheaper to run than `rmake` (one thing to compile rather than 2-3). You'll just need the directives `//@ only-wasm32` and `//@ compile-flags: -Dlinker-messages`.

Please make sure tests have a doc comment describing what they are testing as well.

JayAndJef replied:
Got it. I'll rewrite this next week - is there anything else?

JayAndJef replied:
Just to be clear - this should be a `//@ build-pass` test without an expected output, right?

tgross35 replied:
That sounds correct 👍

Kobzol commented on src/bootstrap/src/core/sanity.rs:
```diff
@@ -314,6 +314,17 @@ than building it.
             .or_insert_with(|| Target::from_triple(&target.triple));
 
+        // compiler-rt c fallbacks for wasm cannot be built with gcc
+        if (target.triple == "wasm32-unknown-unknown" || target.triple == "wasm32v1-none") // bare metal targets without wasi sdk
```
Could you please add a method to `TargetSelection` named e.g. `is_bare_wasm`, or `is_wasm_without_wasi` or something like that that describes this family of targets?

alexcrichton replied:
If going down this compiler-detection route I'd recommend enabling this for all wasm targets, not just the unknown/none targets. The WASI targets have some auto-configuration but the same principle applies here where if gcc is used then that's a bug for WASI as well.

JayAndJef replied:
> If going down this compiler-detection route I'd recommend enabling this for all wasm targets, not just the unknown/none targets.

Would this entail some sort of check to see if the wasi sdk is using gcc?

alexcrichton replied:
In theory (unless I'm misremembering something) nothing more than what's already here other than updating this conditional. The detected compiler for WASI targets should make its way to here and always be clang (wasi-sdk doesn't have gcc and won't for the forseeable future)

Kobzol reviewed (commented):
Detecting GCC would be also useful for other things, but it sounds a bit tricky. Left one nit, otherwise the bootstrap changes (modulo GCC detection, which isn't working yet) look fine. I don't know about WASM though, so someone else should look at that.

jieyouxu commented:
cc @alexcrichton (target maintainer for `wasm32-unknown-unknown`) in case you have any concerns/feedback

tgross35 commented:
> Detecting GCC would be also useful for other things, but it sounds a bit tricky. Left one nit, otherwise the bootstrap changes (modulo GCC detection, which isn't working yet) look fine. I don't know about WASM though, so someone else should look at that.

`cc` exposes this functionality, maybe we could wrap that rather than recreating it here?

jieyouxu commented:
> `cc` exposes this functionality, maybe we could wrap that rather than recreating it here?

Do note that bootstrap `cc` unfortunately has to be pinned to a quite an old version, FWIW.

JayAndJef commented:
Hey yall, just wanted everyone to be aware I'm in a power outage so I probably can't work on this today.

JayAndJef commented:
> There is never any requirement to work on a PR every day, thanks for being active but take your time :)

Thanks, I'll probably finish the changes on Sunday or so then

JayAndJef commented:
I've changed the test to a UI test, and changed the compiler detection. I'm not sure about the consensus on that - are we still going use it? The commits for that can be scrapped if so

alexcrichton commented:
For wasm specifically targets I think that Clang should be used everywhere because gcc doesn't have support for wasm. If any builder is using gcc for wasm that's definitely a bug that needs fixing (which the tests might help detect?)

Kobzol commented:
Well, the dist builders only build stuff, it's not actually tested on most dist builders, which is why https://github.com/rust-lang/rust/issues/132802 was a thing. So the UI/run-make tests are nice to have, but wouldn't help for this situation. But with the bootstrap sanity test I think it's fine.

You can r=me, unless you want to do more changes.

JayAndJef commented:
Great, so the sanity test is going to be kept?

Kobzol commented:
Yeah, let's keep the tests. Let's see if CI works.

@bors try

bors commented:
:hourglass: Trying commit 63db26f1a76bb3c1de04ff6fbc963267effd426c with merge ca07aa7ae640e0f121d5c9af27838d4326b116d5...
<!-- homu: {"type":"TryBuildStarted","head_sha":"63db26f1a76bb3c1de04ff6fbc963267effd426c","merge_sha":"ca07aa7ae640e0f121d5c9af27838d4326b116d5"} -->

rust-log-analyzer commented:
The job **`test-various`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13635523062/job/38113248861) [(plain)](https://github.com/rust-lang-ci/rust/commit/ca07aa7ae640e0f121d5c9af27838d4326b116d5/checks/38113248861/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
Uplifting library (stage1 -> stage2)
[TIMING] core::build_steps::compile::StdLink { compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu }, target_compiler: Compiler { stage: 2, host: x86_64-unknown-linux-gnu }, target: x86_64-unknown-linux-gnu, crates: [], force_recompile: false } -- 0.000
[TIMING] core::build_steps::compile::Std { target: x86_64-unknown-linux-gnu, compiler: Compiler { stage: 2, host: x86_64-unknown-linux-gnu }, crates: [], force_recompile: false, extra_rust_args: [], is_for_mir_opt_tests: false } -- 0.000
##[group]Building test helpers for wasm32-unknown-unknown
cargo:warning=/checkout/tests/auxiliary/rust_test_helpers.c:4:10: fatal error: 'stdlib.h' file not found
cargo:warning=#include <stdlib.h>
cargo:warning=         ^~~~~~~~~~



error occurred: Command "clang-11" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-fno-exceptions" "--target=wasm32-unknown-unknown" "-o" "/checkout/obj/build/wasm32-unknown-unknown/native/rust-test-helpers/cd462aea306bec43-rust_test_helpers.o" "-c" "/checkout/tests/auxiliary/rust_test_helpers.c" with args clang-11 did not execute successfully (status code exit status: 1).

Build completed unsuccessfully in 0:26:04
  local time: Mon Mar  3 17:13:08 UTC 2025
  network time: Mon, 03 Mar 2025 17:13:08 GMT
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13635523062/job/38116440763)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13635523062/job/38116440763","builder_name":"checks-actions"} -->

JayAndJef commented:
oh no

alexcrichton commented:
The failing step there is `Building test helpers for wasm32-unknown-unknown` which I believe is using the C compiler itself to compile `tests/auxiliary/rust_test_helpers.c`. This won't work on wasm32-unknown-unknown since the C compiler has no host headers at all, so that file cannot be compiled. Using clang works as a compiler but it's not a complete toolchain as there's no sysroot for wasm32-unknown-unknown. I think this may be ok for compiler-rt which brings its own (IIRC), but I'm also not certain on this.

jieyouxu commented:
That is used for some UI tests. I suppose, one way to go about that, is to skip building `test_helpers` if the target is wasm32-u-u, and then audit ui test failures to ignore those tests if it's wasm32-u-u(?)

JayAndJef commented:
Would a workaround just be to revert to a run-make test?

JayAndJef commented:
Does anyone see the need for wasm32-unknown-unknown UI tests sometime in the future? If not, I think reverting to a run-make test is a good move

alexcrichton commented:
Personally I'd say no, we probably don't want to test wasm32-unknown-unknown. It's not impossible to test but the lack of debugging facilities (e.g. no println) makes it a non-starter for many workflows. In that sense I'd at least personally be ok with just a run-make test. (or rather I'd be ok with no testing whatsoever and just the bootstrap check, but that's just me)

Kobzol commented:
@bors try

bors commented:
:hourglass: Trying commit 9c25311698c9f415bedee33eff3a7cb2e43b7a3a with merge 8f2e3fb4aa99feaadc334e489e9b781e67101534...
<!-- homu: {"type":"TryBuildStarted","head_sha":"9c25311698c9f415bedee33eff3a7cb2e43b7a3a","merge_sha":"8f2e3fb4aa99feaadc334e489e9b781e67101534"} -->

rust-log-analyzer commented:
The job **`test-various`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13715420595/job/38359358927) [(plain)](https://github.com/rust-lang-ci/rust/commit/8f2e3fb4aa99feaadc334e489e9b781e67101534/checks/38359358927/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
Uplifting library (stage1 -> stage2)
[TIMING] core::build_steps::compile::StdLink { compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu, forced_compiler: true }, target_compiler: Compiler { stage: 2, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu, crates: [], force_recompile: false } -- 0.000
[TIMING] core::build_steps::compile::Std { target: x86_64-unknown-linux-gnu, compiler: Compiler { stage: 2, host: x86_64-unknown-linux-gnu, forced_compiler: false }, crates: [], force_recompile: false, extra_rust_args: [], is_for_mir_opt_tests: false } -- 0.000
##[group]Building test helpers for wasm32-unknown-unknown
cargo:warning=/checkout/tests/auxiliary/rust_test_helpers.c:4:10: fatal error: 'stdlib.h' file not found
cargo:warning=#include <stdlib.h>
cargo:warning=         ^~~~~~~~~~



error occurred: Command "clang-11" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-fno-exceptions" "--target=wasm32-unknown-unknown" "-o" "/checkout/obj/build/wasm32-unknown-unknown/native/rust-test-helpers/cd462aea306bec43-rust_test_helpers.o" "-c" "/checkout/tests/auxiliary/rust_test_helpers.c" with args clang-11 did not execute successfully (status code exit status: 1).

Build completed unsuccessfully in 0:25:35
  local time: Fri Mar  7 07:44:25 UTC 2025
  network time: Fri, 07 Mar 2025 07:44:25 GMT
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13715420595/job/38361178763)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13715420595/job/38361178763","builder_name":"checks-actions"} -->

JayAndJef commented:
Strange, worked when I ran it on my machine. Will look into later

JayAndJef commented:
Removed the test, is that ok with everyone

Kobzol commented on src/bootstrap/src/core/sanity.rs:
```diff
@@ -376,3 +387,16 @@ $ pacman -R cmake && pacman -S mingw-w64-x86_64-cmake
+        .arg("-x")
+        .arg("c")
+        .arg("/dev/null")
+        .run_capture_stdout(build)
```
It seems like GCC supports these flags, but maybe some other compilers won't (in particular MSVC?). I guess that if the command fails we should return false from the function rather than failing the build. There is a method for allowing failure on the command.

JayAndJef replied:
👍

jyn514 replied:
is there any reason you're not using `cc::Tool::is_like_clang`, as has already been suggested above?

JayAndJef replied:
> is there any reason you're not using `cc::Tool::is_like_clang`, as has already been suggested above?

I wasn't sure on it's behavior. Is there a way to construct a `Tool` from the path given in config?

Kobzol replied:
The `Build::cc` function that you already use accesses `cc::Tool` internally. It just now returns a `PathBuf`, and doesn't provide access to the inner `Tool`. You can add another function, e.g. `cc_tool`, that will return a reference to the tool directly, and then call `is_like_clang` on it.

Kobzol commented:
Thanks. Anything else you'd like to do in this PR?

JayAndJef commented:
> Thanks. Anything else you'd like to do in this PR?

I don't think so

Kobzol commented:
@bors try

bors commented:
:hourglass: Trying commit faddfc192bd863f68ceae746e0f7a2d0bbe9fb74 with merge 085d3499532e2e76cdc834a2f12e7b5df0a520d3...
<!-- homu: {"type":"TryBuildStarted","head_sha":"faddfc192bd863f68ceae746e0f7a2d0bbe9fb74","merge_sha":"085d3499532e2e76cdc834a2f12e7b5df0a520d3"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13791505936/job/38583513070)
Build commit: 085d3499532e2e76cdc834a2f12e7b5df0a520d3 (`085d3499532e2e76cdc834a2f12e7b5df0a520d3`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13791505936/job/38583513070"},"merge_sha":"085d3499532e2e76cdc834a2f12e7b5df0a520d3"} -->

Kobzol commented:
@bors r+

bors commented:
:pushpin: Commit faddfc192bd863f68ceae746e0f7a2d0bbe9fb74 has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol faddfc192bd863f68ceae746e0f7a2d0bbe9fb74 -->
<!-- homu: {"type":"Approved","sha":"faddfc192bd863f68ceae746e0f7a2d0bbe9fb74","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

tgross35 commented:
This PR has 13 commits with a lot of back and forth @JayAndJef could you squash?

@bors r- (sorry Jakub)

Kobzol commented:
Sorry, I saw that most changes were deleted and kind of assumed the commits were gone too :sweat_smile: Thanks for noticing.

JayAndJef commented:
How do I squash?

tgross35 commented:
Short answer: `git rebase -i master --keep-base`, replace "pick" with "s" (or "squash") for everything except the first line, delete the outdated parts of the message and write a good summary+description of what the final version of the change does, save & exit, then `git push --force-with-lease` to update your branch.

There are some more detailed instructions here https://rustc-dev-guide.rust-lang.org/git.html#squash-your-commits

Kobzol commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit e1bb12dff2c91c71df9895e03524e02ffeebf592 has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol e1bb12dff2c91c71df9895e03524e02ffeebf592 -->
<!-- homu: {"type":"Approved","sha":"e1bb12dff2c91c71df9895e03524e02ffeebf592","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r- rollup=iffy
https://github.com/rust-lang/rust/pull/138443#issuecomment-2720614237

JayAndJef commented:
What happened?

Kobzol commented:
The `dist-various-1` job has failed with this PR. You can see the failure [here](https://github.com/rust-lang-ci/rust/actions/runs/13830402120/job/38693244518).

```
2025-03-13T09:44:03.1078299Z [TIMING] core::build_steps::test::Compiletest { compiler: Compiler { stage: 2, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: thumbv7em-none-eabihf, mode: "run-make", suite: "run-make", path: "tests/run-make", compare_mode: None } -- 91.046
2025-03-13T09:44:03.1080605Z [TIMING] core::build_steps::test::RunMake { compiler: Compiler { stage: 2, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: thumbv7em-none-eabihf } -- 0.000
2025-03-13T09:44:03.1134493Z Build completed successfully in 0:53:14
2025-03-13T09:44:03.1225863Z + python3 ../x.py dist --host= --target wasm32-unknown-emscripten,arm-unknown-linux-musleabi,arm-unknown-linux-musleabihf,armv5te-unknown-linux-gnueabi,armv5te-unknown-linux-musleabi,armv7-unknown-linux-musleabihf,aarch64-unknown-none,aarch64-unknown-none-softfloat,sparc64-unknown-linux-gnu,x86_64-unknown-redox,thumbv6m-none-eabi,thumbv7m-none-eabi,thumbv7em-none-eabi,thumbv7em-none-eabihf,thumbv8m.base-none-eabi,thumbv8m.main-none-eabi,thumbv8m.main-none-eabihf,riscv32i-unknown-none-elf,riscv32im-unknown-none-elf,riscv32imc-unknown-none-elf,riscv32imac-unknown-none-elf,riscv32imafc-unknown-none-elf,riscv64imac-unknown-none-elf,riscv64gc-unknown-none-elf,armebv7r-none-eabi,armebv7r-none-eabihf,armv7r-none-eabi,armv7r-none-eabihf,thumbv7neon-unknown-linux-gnueabihf,armv7a-none-eabi,aarch64-pc-windows-gnullvm,i686-pc-windows-gnullvm,x86_64-pc-windows-gnullvm
2025-03-13T09:44:03.1769182Z ##[group]Building bootstrap
2025-03-13T09:44:03.2908262Z     Finished `dev` profile [unoptimized] target(s) in 0.11s
2025-03-13T09:44:03.2986926Z ##[endgroup]
2025-03-13T09:44:03.4044673Z cargo:warning=Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `sparc64-linux-gnu-g++` installed?
2025-03-13T09:44:03.4050342Z cargo:warning=Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `sparc64-linux-gnu-g++` installed?
2025-03-13T09:44:04.3085304Z cargo:warning=Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `arm-linux-musleabi-g++` installed?
2025-03-13T09:44:04.3092256Z cargo:warning=Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `arm-linux-musleabi-g++` installed?
2025-03-13T09:44:04.6082582Z cargo:warning=Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `arm-linux-musleabihf-g++` installed?
2025-03-13T09:44:04.6088903Z cargo:warning=Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `arm-linux-musleabihf-g++` installed?
2025-03-13T09:44:04.7963599Z cargo:warning=Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `emcc` installed?
2025-03-13T09:44:04.7973991Z cargo:warning=Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `em++` installed?
2025-03-13T09:44:04.7984445Z cargo:warning=Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `em++` installed?
2025-03-13T09:44:04.9649878Z cargo:warning=Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `arm-linux-musleabihf-g++` installed?
2025-03-13T09:44:04.9656382Z cargo:warning=Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `arm-linux-musleabihf-g++` installed?
2025-03-13T09:44:04.9805961Z 
2025-03-13T09:44:04.9806671Z thread 'main' panicked at src/bootstrap/src/core/sanity.rs:321:17:
2025-03-13T09:44:04.9807249Z only clang supports building c code for wasm targets
2025-03-13T09:44:04.9807754Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Probably we also use GCC for building wasm targets on `dist-various-1`?

JayAndJef commented:
got it, i thought dist various 1 didnt have wasm targets

JayAndJef committed:
fix: install correct cc for wasm32-unknown-emscripten

Also fixed a typo in the sanity check for bootstrap, as we are checking for clang-likeness in every wasm target.

Kobzol commented:
@bors try

bors commented:
:lock: Merge conflict

This pull request and the master branch diverged in a way that cannot be automatically merged. Please rebase on top of the latest master branch, and let the reviewer approve again.

<details><summary>How do I rebase?</summary>

Assuming `self` is your fork and `upstream` is this repository, you can resolve the conflict following these steps:

1. `git checkout issue-132802-fix` *(switch to your branch)*
2. `git fetch upstream master` *(retrieve the latest master)*
3. `git rebase upstream/master -p` *(rebase on top of it)*
4. Follow the on-screen instruction to resolve conflicts (check `git status` if you got lost).
5. `git push self issue-132802-fix --force-with-lease` *(update this PR)*

You may also read [*Git Rebasing to Resolve Conflicts* by Drew Blessing](http://blessing.io/git/git-rebase/open-source/2015/08/23/git-rebasing-to-resolve-conflicts.html) for a short tutorial.

Please avoid the ["**Resolve conflicts**" button](https://help.github.com/articles/resolving-a-merge-conflict-on-github/) on GitHub. It uses `git merge` instead of `git rebase` which makes the PR commit history more difficult to read.

Sometimes step 4 will complete without asking for resolution. This is usually due to difference between how `Cargo.lock` conflict is handled during merge and rebase. This is normal, and you should still perform step 5 to update this PR.

</details>

<details><summary>Error message</summary>

```text
Auto-merging src/bootstrap/src/lib.rs
CONFLICT (content): Merge conflict in src/bootstrap/src/lib.rs
Auto-merging src/bootstrap/src/core/sanity.rs
Automatic merge failed; fix conflicts and then commit the result.

```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #138523) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Kobzol commented:
@bors try

bors commented:
:hourglass: Trying commit dfe4ac0ba7edecae08f49973eb25ab40236bdfcc with merge 207f62cb4f2f32103bbde3d0110f385abd73e6b2...
<!-- homu: {"type":"TryBuildStarted","head_sha":"dfe4ac0ba7edecae08f49973eb25ab40236bdfcc","merge_sha":"207f62cb4f2f32103bbde3d0110f385abd73e6b2"} -->

rust-log-analyzer commented:
The job **`dist-various-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13886608341/job/38852164696) [(plain)](https://github.com/rust-lang-ci/rust/commit/207f62cb4f2f32103bbde3d0110f385abd73e6b2/checks/38852164696/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.url=https://github.com/rust-lang/rust
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---

COPY host-x86_64/dist-various-1/install-emscripten.sh /build
RUN ./install-emscripten.sh

# Add Emscripten to PATH
ENV PATH="/build/emsdk:/build/emsdk/upstream/emscripten:/build/emsdk/node/current/bin:${PATH}"

# Suppress some warnings in the openwrt toolchains we downloaded
ENV STAGING_DIR=/tmp

COPY scripts/musl.sh /build
---
#23 2.147   cups-common liblcms2-utils pciutils lm-sensors npm libnss-mdns
#23 2.147   fonts-dejavu-extra fonts-ipafont-gothic fonts-ipafont-mincho
#23 2.147   fonts-wqy-microhei | fonts-wqy-zenhei fonts-indic
#23 2.147 Recommended packages:
#23 2.147   libgl1-amber-dri nodejs-doc libatk-wrapper-java-jni fonts-dejavu-extra
#23 2.213 The following NEW packages will be installed:
#23 2.213   ca-certificates-java default-jre default-jre-headless fontconfig-config
#23 2.213   fonts-dejavu-core java-common libavahi-client3 libavahi-common-data
#23 2.213   libavahi-common3 libc-ares2 libcups2 libdrm-amdgpu1 libdrm-intel1
#23 2.213   libdrm-nouveau2 libdrm-radeon1 libedit2 libfontconfig1 libfreetype6 libgif7
---
#23 7.836 
#23 8.008 done.
#23 8.008 done.
#23 8.045 + git clone https://github.com/emscripten-core/emsdk.git
#23 8.047 Cloning into 'emsdk'...
#23 8.448 + cd emsdk
#23 8.448 + ./emsdk install latest
#23 8.654 Resolving SDK alias 'latest' to '4.0.5'
#23 8.654 Resolving SDK version '4.0.5' to 'sdk-releases-d7f8ff5e2ca3539c33fae81e98f7c56ef9fa1239-64bit'
#23 8.654 Installing SDK 'sdk-releases-d7f8ff5e2ca3539c33fae81e98f7c56ef9fa1239-64bit'..
#23 8.654 Installing tool 'node-20.18.0-64bit'..
#23 8.654 Downloading: /build/emsdk/downloads/node-v20.18.0-linux-x64.tar.xz from https://storage.googleapis.com/webassembly/emscripten-releases-builds/deps/node-v20.18.0-linux-x64.tar.xz, 25738964 Bytes
#23 8.654  [----------------------------------------------------------------------------]
#23 10.83 Unpacking '/build/emsdk/downloads/node-v20.18.0-linux-x64.tar.xz' to '/build/emsdk/node/20.18.0_64bit'
#23 10.83 Done installing tool 'node-20.18.0-64bit'.
#23 10.83 Installing tool 'releases-d7f8ff5e2ca3539c33fae81e98f7c56ef9fa1239-64bit'..
#23 10.83 Downloading: /build/emsdk/downloads/d7f8ff5e2ca3539c33fae81e98f7c56ef9fa1239-wasm-binaries.tar.xz from https://storage.googleapis.com/webassembly/emscripten-releases-builds/linux/d7f8ff5e2ca3539c33fae81e98f7c56ef9fa1239/wasm-binaries.tar.xz, 300524536 Bytes
#23 10.83  [----------------------------------------------------------------------------]
#23 31.64 Unpacking '/build/emsdk/downloads/d7f8ff5e2ca3539c33fae81e98f7c56ef9fa1239-wasm-binaries.tar.xz' to '/build/emsdk/upstream'
#23 31.64 Done installing tool 'releases-d7f8ff5e2ca3539c33fae81e98f7c56ef9fa1239-64bit'.
#23 31.64 Done installing SDK 'sdk-releases-d7f8ff5e2ca3539c33fae81e98f7c56ef9fa1239-64bit'.
#23 31.65 + ./emsdk activate latest
#23 31.74 Resolving SDK alias 'latest' to '4.0.5'
#23 31.74 Resolving SDK version '4.0.5' to 'sdk-releases-d7f8ff5e2ca3539c33fae81e98f7c56ef9fa1239-64bit'
#23 31.74 Setting the following tools as active:
#23 31.74    node-20.18.0-64bit
#23 31.74    releases-d7f8ff5e2ca3539c33fae81e98f7c56ef9fa1239-64bit
#23 31.74 
#23 31.74 Next steps:
---
cargo:warning=Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `arm-linux-musleabihf-g++` installed?
cargo:warning=Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `arm-linux-musleabihf-g++` installed?
cargo:warning=Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `sparc64-linux-gnu-g++` installed?
cargo:warning=Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `sparc64-linux-gnu-g++` installed?
cargo:warning=Compiler family detection failed due to error: ToolExecError: Command "emcc" "-E" "/tmp/12388544568563907636detect_compiler_family.c" with args emcc did not execute successfully (status code exit status: 1).
cargo:warning=Compiler family detection failed due to error: ToolExecError: Command "em++" "-E" "/tmp/5009887824444102905detect_compiler_family.c" with args em++ did not execute successfully (status code exit status: 1).
cargo:warning=Compiler family detection failed due to error: ToolExecError: Command "em++" "-E" "/tmp/15472697614068236456detect_compiler_family.c" with args em++ did not execute successfully (status code exit status: 1).
cargo:warning=Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `arm-linux-musleabihf-g++` installed?
cargo:warning=Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `arm-linux-musleabihf-g++` installed?

thread 'main' panicked at src/bootstrap/src/core/sanity.rs:321:17:
only clang supports building c code for wasm targets
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13886608341/job/38853325073)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13886608341/job/38853325073","builder_name":"checks-actions"} -->

alex-semenyuk commented:
@JayAndJef
Thanks for your contribution
From wg-triage. Any updates on this PR?

bors commented:
:umbrella: The latest upstream changes (presumably #140772) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tgross35 commented:
@JayAndJef any chance you are able to pick this up? You probably need to install `clang` in the dockerfiles.

JayAndJef commented:
Sorry yall, I kind of completely forgot about this pr. Is anyone else willing to take this on?

tgross35 commented:
Well, somebody did mention interest in https://github.com/rust-lang/rust/issues/132802. But this seems close enough to get over the line if you are able! Needs a rebase plus Clang should be installed, probably not too much after that.

tgross35 commented:
No worries! It's open source and life happens :)

JayAndJef commented:
Sorry, what should I do so that the changes would be mergeable?

tgross35 commented on src/bootstrap/src/core/sanity.rs:
```diff
@@ -311,6 +311,17 @@ than building it.
             .or_insert_with(|| Target::from_triple(&target.triple));
 
+        // compiler-rt c fallbacks for wasm cannot be built with gcc
+        if target.contains("wasm")
+            && (build.config.optimized_compiler_builtins(*target)
+                || build.config.rust_std_features.contains("compiler-builtins-c"))
+        {
+            let is_clang = build.cc_tool(*target).is_like_clang();
+            if !is_clang {
+                panic!("only clang supports building c code for wasm targets");
+            }
+        }
```
Mind adding some more guidance to the panic message so anyone who hits this doesn't have to dig too much? Maybe something like

```rust
                panic!(
                    "Clang is required to build C code for Wasm targets, got `{}` instead\n\
                    this is because compiler-builtins is configured to build C source. Either \
                    ensure Clang is used, or adjust this configuration.",
                    tool.path().display()
                );
```

JayAndJef commented:
I keep getting merge conflicts while rebasing and I don't know how to solve them.

```
warning: skipped previously applied commit aa67f0ca2a3
hint: use --reapply-cherry-picks to include skipped commits
hint: Disable this message with "git config set advice.skippedCherryPicks false"
Auto-merging Cargo.lock
Auto-merging src/bootstrap/src/core/build_steps/llvm.rs
Failed to merge submodule src/tools/cargo (commits not present)
CONFLICT (submodule): Merge conflict in src/tools/cargo
hint: Recursive merging with submodules currently only supports trivial cases.
hint: Please manually handle the merging of each conflicted submodule.
hint: This can be accomplished with the following steps:
hint:  - come back to superproject and run:
hint:
hint:       git add src/tools/cargo
hint:
hint:    to record the above merge or update
hint:  - resolve any other conflicts in the superproject
hint:  - commit the resulting index in the superproject
hint:
hint: Disable this message with "git config set advice.submoduleMergeConflict false"
error: could not apply 6efacfb7a59... add fix for full tools and sanitizer
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
hint: Disable this message with "git config set advice.mergeConflict false"
Could not apply 6efacfb7a59... add fix for full tools and sanitizer
```

I tried manually updating the cargo submodule to the latest commit in upstream, but then when I run git add and continue the rebase it complains that I haven't solved the conflict and added the changes.

JayAndJef commented:
I'm incredibly confused on wtf is going on with git right now because I keep running into unexpected things - I'm going to take a break and hopefully find some clarity tomorrow.

Just to be clear, I was supposed to sync the repository then `git pull --rebase`?

tgross35 commented:
What are you running into?

I think the Get Out of Jail Free card should look something like this:

```sh
# Ensure you are on the right branch and it is up to date
git switch issue-132802-fix
git pull --rebase
# ^ if that fails for whatever reason, you can do a more extreme
# `git reset --hard dfe4ac0ba7ed`, which will blow away any chances and
# reset your branch to the last commit on this PR. Careful!

# Forget submodules
git submodule deinit --all --force

# Fetch rust-lang/rust. The latest commit on master will be FETCH_HEAD
git fetch https://github.com/rust-lang/rust.git master

# Do the actual rebase. You will need to resolve a conflict; do that, `git add .`,
# and `git rebase --continue`.
#
# Make sure you don't accidentally add a submodule here, sometimes git can be
# weird like that. If you see anything suspicious, `git submodule deinit --all --force`
git rebase FETCH_HEAD

# And update the PR
git push --force-with-lease
```

JayAndJef commented on src/ci/docker/host-x86_64/dist-various-1/Dockerfile:
```diff
@@ -55,6 +55,15 @@ RUN ./install-riscv64-none-elf.sh
 RUN ./install-riscv32-none-elf.sh
 
+COPY host-x86_64/dist-various-1/install-llvm-mingw.sh /build
+RUN ./install-llvm-mingw.sh
```
I'm not sure why this was here, removing it

JayAndJef committed:
fix: error message

JayAndJef committed:
fix: remove unneeded(?) install script

Kobzol commented:
@bors2 try

tgross35 commented:
Good point, `cc` should be using `emcc` by default. I guess maybe fix the current error first, then we can rerun tests? Now that it prints `cc_tool.path().display()`, we should get a bit more info.

JayAndJef commented:
test-various works when ran locally, and it seems like the error was that the gnu mirror kept timing out. Maybe it was a one-off network issue when running CI? 

```
#13 329.5 failed: Connection timed out.
#13 329.5 Connecting to gitweb.git.savannah.gnu.org (gitweb.git.savannah.gnu.org)|135.148.138.35|:443... failed: Connection timed out.
#13 329.5 Connecting to gitweb.git.savannah.gnu.org (gitweb.git.savannah.gnu.org)|38.102.85.135|:443... connected.
--- ERROR ---
#13 329.5 HTTP request sent, awaiting response... 502 Bad Gateway
#13 329.5 2025-07-01 21:42:40 ERROR 502: Bad Gateway.
#13 329.5
#13 329.5 make: *** [Makefile:79: sources/config.sub] Error 8
--- ERROR ---
#13 ERROR: process "/bin/sh -c bash musl-toolchain.sh x86_64 && rm -rf build" did not complete successfully: exit code: 1
```

tgross35 commented:
@bors2 try

JayAndJef commented:
Will look into soon

tgross35 commented:
For lack of a better idea, let's just see where the other jobs get

@bors2 try jobs=dist-various*

tgross35 commented:
Seems like the change to the dockerfile just busts cache and we're hitting a spurious download. Discussion at [#t-infra > test-various](https://rust-lang.zulipchat.com/#narrow/channel/242791-t-infra/topic/test-various/with/526940207)

JayAndJef committed:
fix: allow emcc

JayAndJef commented:
makes me wonder if there's a better way to detect emcc specifically, or group it into clang

tgross35 commented on src/ci/docker/host-x86_64/test-various/Dockerfile:
```diff
@@ -59,8 +60,8 @@ RUN curl -L https://github.com/bytecodealliance/wasmtime/releases/download/v19.0
 ENV PATH "$PATH:/wasmtime-v19.0.0-x86_64-linux"
 
-ENV WASM_TARGETS=wasm32-wasip1
-ENV WASM_SCRIPT python3 /checkout/x.py --stage 2 test --host='' --target $WASM_TARGETS \
+ENV WASM_WASIP_TARGET=wasm32-wasip1 
+ENV WASM_WASIP_SCRIPT python3 /checkout/x.py --stage 2 test --host='' --target $WASM_WASIP_TARGET \
```
Actually, why is this wasi build succeeding? We don't set `CC_wasm32_wasip1` to Clang anywhere, which would mean GCC is the default.

I think it would be good to set the `CC_*` variables even if it works for some reason, to make it explicit that we want clang.

JayAndJef replied:
> What exactly is going on with the changes in this file btw? It looks like a variable rename, which should be inconsequential, plus installing gcc-multilib?

I'm not very sure about the variable changes... maybe I renamed it for clarity?
If I remember correctly installing gcc-multilib was in response to the script failing locally.

> Actually, why is this wasi build succeeding? We don't set `CC_wasm32_wasip1` to Clang anywhere, which would mean GCC is the default.
> 
> I think it would be good to set the `CC_*` variables even if it works for some reason, to make it explicit that we want clang.

`wasm32-wasip1` depends on the wasi sdk, which is based on clang. You can see it installed on line 44-46 of the `test-various` Dockerfile.

Kobzol commented:
I haven't followed all the discussion here - do we *have* to use emscripten and cannot use clang?

JayAndJef commented:
> I haven't followed all the discussion here - do we *have* to use emscripten and cannot use clang?

I'm pretty sure so, but to be safe maybe someone who works on that target could answer?

Kobzol commented on src/bootstrap/src/core/sanity.rs:
```diff
@@ -328,6 +328,23 @@ than building it.
+                || build.config.rust_std_features.contains("compiler-builtins-c"))
+        {
+            let cc_tool = build.cc_tool(*target);
+            if !cc_tool.is_like_clang() && !cc_tool.path().ends_with("emcc") {
```
That seems unrelated to this PR. Maybe the recent `cargo update` bump?

ognevny replied:
oh, I forgot to up-to-date rustc-perf checkout. I switched to tarball source and now it works, sorry for noise...
